panel/io/cache.py

```diff
@@ -0,0 +1,307 @@
+"""
+Implements memoization for functions with arbitrary arguments
+"""
+import collections
+import functools
+import hashlib
+import inspect
+import io
+import pickle
+import threading
+import time
+import unittest
+import weakref
+
+import unittest.mock
+
+import param
+
+from .state import state
+
+#---------------------------------------------------------------------
+# Private API
+#---------------------------------------------------------------------
+
+_CYCLE_PLACEHOLDER = b"panel-93KZ39Q-floatingdangeroushomechose-CYCLE"
+
+_NATIVE_TYPES = (
+    bytes, str, float, int, bool, bytearray, type(None)
+)
+
+_FFI_TYPE_NAMES = ("_cffi_backend.FFI", "builtins.CompiledFFI",)
+
+_HASH_MAP = dict()
+
+_HASH_STACKS = weakref.WeakKeyDictionary()
+
+_INDETERMINATE = type('INDETERMINATE', (object,), {})()
+
+_TIME_FN = time.monotonic
+
+class _Stack(object):
+
+    def __init__(self):
+        self._stack = collections.OrderedDict()
+
+    def push(self, val):
+        self._stack[id(val)] = val
+
+    def pop(self):
+        self._stack.popitem()
+
+    def __contains__(self, val):
+        return id(val) in self._stack
+
+def _get_fqn(obj):
+    """Get module.type_name for a given type."""
+    the_type = type(obj)
+    module = the_type.__module__
+    name = the_type.__qualname__
+    return "%s.%s" % (module, name)
+
+def _int_to_bytes(i):
+    num_bytes = (i.bit_length() + 8) // 8
+    return i.to_bytes(num_bytes, "little", signed=True)
+
+def _is_native(obj):
+    return isinstance(obj, _NATIVE_TYPES)
+
+def _is_native_tuple(obj):
+    return isinstance(obj, tuple) and all(_is_native_tuple(v) for v in obj)
+
+def _container_hash(obj):
+    h = hashlib.new("md5")
+    h.update(_generate_hash(f'__{type(obj).__name__}'))
+    for item in (obj.items() if isinstance(obj, dict) else obj):
+        h.update(_generate_hash(item))
+    return h.digest()
+
+def _partial_hash(obj):
+    h = hashlib.new("md5")
+    h.update(_generate_hash(obj.args))
+    h.update(_generate_hash(obj.func))
+    h.update(_generate_hash(obj.keywords))
+    return h.digest()
+
+def _pandas_hash(obj):
```

Comment:
Copied from streamlit, not entirely happy with this since I don't consider hashing on a sample to be safe.